providers/file/file.go

```diff
@@ -122,7 +123,19 @@ func (f *File) Watch(cb func(event interface{}, err error)) error {
 
 		w.Close()
 	}()
-
+	f.watchers = append(f.watchers, w)
 	// Watch the directory for changes.
 	return w.Add(fDir)
 }
+
+// Unwatch stops watching the files. It closes all the fsnotify watchers
+// and event channels associated with them.
+func (f *File) Unwatch() error {
+	for _, watcher := range f.watchers {
```

Comment:
Does f.watchers need to be reset to an empty list after calling Unwatch? Also can there be more than one watcher?

Reply:
Hmm... I'm not sure what will be the use of having multiple watchers on the same file and calling `.Watch` twice on the same provider as the file path can't be changed, correct?

I would rather assume that calling .Watch() again would just replace the watcher inside the provider. Maybe it might be better to use and store a single fsnotify Watcher.

Reply:
@rhnvrm, there could be a scenario where multiple watchers exist instead of being nested inside the callback function. I was experimenting with one of my previous [commit](https://github.com/knadh/koanf/pull/245/commits/313886424a5e956f8cc015a9bcdd6f5ad703a408), which was exactly what you are suggesting. In this case, only one watcher (in random order) is being called when the file is changed.

try adding this code snippet to my previous example after checking out the commit I mentioned.
```go
	i := 0
	go func() {
		for true {
			time.Sleep(3 * time.Second)
			i++
			fmt.Println("writing to file")
			os.WriteFile(fileName, []byte(`{"parent": {"name": "name"}}`), 0600)
			fmt.Println("i = ", i)
			if i == 3 {
				fmt.Println("Unwatching")
				err := f.Unwatch()
				if err != nil {
					return
				}
			}
		}
	}()
```

Output
```
2023/12/05 16:41:21 watcher 2 config changed. Reloading ...
writing to file
i =  1
2023/12/05 16:41:23 watcher 1 config changed. Reloading ...
writing to file
i =  2
2023/12/05 16:41:26 watcher 2 config changed. Reloading ...
writing to file
2023/12/05 16:41:29 watcher 2 config changed. Reloading ...
i =  3
Unwatching
2023/12/05 16:41:29 watch error: fsnotify watch channel closed
2023/12/05 16:41:29 watch error: fsnotify watch channel closed
```

PS: i had modified the commit's `f.Unwatch` function and closed the watcher using `f.w.Close()`

Reply:
@rhnvrm I have updated the code to store only a single fsnotify watcher.